model/util.py

```diff
@@ -0,0 +1,91 @@
+import math
+
+
+class TemporalFilter:
+
+    def __init__(self, tau_rise: float, tau_fall: float) -> None:
+        """
+        tau_rise:
+         * This controls how quickly the filter responds to an increase in the
+           prediction error.
+         * A smaller tau_rise means a faster response to changes, making the
+           filter more sensitive to sudden increases in error.
+         * A larger tau_rise will smooth out the response, making the filter
+           less sensitive to abrupt changes.
+
+        tau_fall:
+         * This parameter determines the decay rate of the filter's response.
+         * A smaller tau_fall will make the filter's response to decreases in
+           the error signal more rapid.
+         * A larger tau_fall will result in a slower decay, allowing the filter
+           to retain the influence of past errors for longer.
+
+        If you're dealing with high-frequency noise, you might start with a
+        tau_rise that's quite small (fractions of the time step) to quickly
+        adapt to changes, while tau_fall could be larger to prevent the filter
+        from reacting too strongly to what might be noise. Conversely, for a
+        system with slow-moving dynamics, you might choose larger values for
+        both tau_rise and tau_fall to avoid reacting to insignificant changes.
+        """
+        self.rise: float = 0
+        self.fall: float = 0
+        self.tau_rise = tau_rise
+        self.tau_fall = tau_fall
+
+    def apply(self, error: float, dt: float = 1) -> float:
```

Comment:
This is looking great. One minor suggestion, nice to hint at units on dt and taurise and taufall if we intend to match units from literature.

Reply:
Yeah I can incorporate this into a later diff here where I am touching these files again.
https://github.com/and-rewsmith/LPL-SNN/pull/23